### ISSUE TITLE
Clang 18 does not have flag `-fno-objc-msgsend-selector-stubs`

### ISSUE THREAD

**Issue by DUOLabs333** [NixOS/nixpkgs#343418]
### Describe the bug
See title --- trying to built with that flag fails

### Steps To Reproduce
1. Run `clang++ -fno-objc-msgsend-selector-stubs`

### Expected behavior
Clang gained support for this flag with LLVM 16, so I expected the flag to exist by 18.

### Notify maintainers
I can't find the list for clang-18 in particular

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"aarch64-darwin"`
 - host os: `Darwin 20.4.0, macOS 11.3.1`
 - multi-user?: `yes`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.18.1`
 - nixpkgs: `/Users/system/.nix-defexpr/channels/nixpkgs`
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

**Comment by RossComputerGuy** [comment]
> I can't find the list for clang-18 in particular

@NixOS/llvm 

> Steps to reproduce the behavior:
> 
>     1. Run `clang++ -fno-objc-msgsend-selector-stubs`

Which clang derivation did you use?

**Comment by alyssais** [comment]
Does this flag definitely exist in upstream Clang, as opposed to just in Apple's fork? I can't find any mention of it in my LLVM checkout.

**Comment by DUOLabs333** [comment]
@alyssais [This issue](https://github.com/llvm/llvm-project/issues/56034) tracked the flag's addition to LLVM 16.

**Comment by DUOLabs333** [comment]
@RossComputerGuy `zicyagjyqqmkvqcljcpl5blhfrab28mr-clang-wrapper-18.1.7`

**Comment by RossComputerGuy** [comment]
@DUOLabs333 Please don't just send a store directory name, that doesn't help if people don't use the same platform you use. Instead, share the attribute name inside of `pkgs` which you've used.

**Comment by DUOLabs333** [comment]
@RossComputerGuy Oh, I wasn't sure exactly what you mean. I used `pkgs.clang-18`.

**Comment by RossComputerGuy** [comment]
> Does this flag definitely exist in upstream Clang, as opposed to just in Apple's fork? I can't find any mention of it in my LLVM checkout.

I don't see it in the llvmorg-16.0.6 branch of my LLVM checkout as well. It likely could be a Darwin only flag in their fork?

**Comment by RossComputerGuy** [comment]
> Oh, I wasn't sure exactly what you meant. I used `pkgs.clang-18`.

Thank you, it looks like LLVM 18 doesn't have this flag either.

**Comment by RossComputerGuy** [comment]
I've gone ahead and asked in LLVM's Discord server if this is only in Apple's fork. Also asked the NixOS Darwin maintainers if that's the case as well.

**Comment by RossComputerGuy** [comment]
It appears Apple's fork did contain it at some point, SDK 24 does have it. Unfortunately, this comes from Apple's private fork so we won't be able to add this flag in.

**Comment by DUOLabs333** [comment]
Hmmm, I'm not sure what that issue was referring to then.

**Comment by reckenrode** [comment]
That issue was implementing compatibility in LLD with the optimization that Apple started doing in Xcode 14. I don’t know whether LLVM’s clang performs a similar optimization or not. It doesn’t seem to have a flag controlling it.

nixpkgs builds ld64 951.9, which is the version that is shipped with Xcode 15.3. It should be compatible with the optimization that necessitated that flag if you’re trying to link code that uses it.

**Comment by DUOLabs333** [comment]
Ah, thanks for the clarification.
